[Foundation] Address exclusivity violation in Data

devincoughlin opened the pull request:
Address an exclusivity violation in Data's Iterator.next() by changing two private
'let' stored properties to be 'var'.

Making the properties 'var' changes code generation of next() so that the stored
properties are read independently of the other contents of the struct. This prevents
an exclusivity violation when reading '_endIndex' and '_data' while simultaneously
mutating '_buffer' with the call to withUnsafeMutablePointer().

The 'let' pattern is an idiom we would eventually like to support, but for now we
need to remove the exclusivity violation.

devincoughlin commented:
@swift-ci Please test

jrose-apple commented:
Can you leave behind a comment about this? Normally the only reason to make properties in a struct use `let` is because there are invariants across multiple properties, and I think that is in fact the case here.

gottesmm commented:
Also, probably needs some sort of in text reference to the bug to undo this.

devincoughlin committed:
[Foundation] Address exclusivity violation in Data

Address an exclusivity violation in Data's Iterator.next() by changing two private
'let' stored properties to be 'var'.

Making the properties 'var' changes code generation of next() so that the stored
properties are read independently of the other contents of the struct. This
prevents an exclusivity violation when reading '_endIdx' and '_data' while
simultaneously mutating '_buffer' with the call to withUnsafeMutablePointer().

The 'let' pattern is an idiom we would eventually like to support (see SR-7396),
but for now we need to remove the exclusivity violation.

devincoughlin commented:
@swift-ci please smoke test and merge

devincoughlin commented:
@swift-ci please smoke test and merge
